src/Adapter/Shop/MaintenanceConfiguration.php

```diff
@@ -61,24 +68,60 @@ public function getConfiguration()
      */
     public function updateConfiguration(array $configuration)
     {
-        if ($this->validateConfiguration($configuration)) {
-            $this->configuration->set('PS_SHOP_ENABLE', $configuration['enable_shop']);
-            $this->configuration->set('PS_MAINTENANCE_IP', $configuration['maintenance_ip']);
-            $this->configuration->set('PS_MAINTENANCE_TEXT', $configuration['maintenance_text'], null, ['html' => true]);
+        $shopConstraint = null;
+        if (!$this->shopContext->isAllShopContext()) {
+            $configuration = $this->removeDisabledFields($configuration);
+            $contextShopGroup = $this->shopContext->getContextShopGroup();
+            $contextShopId = $this->shopContext->getContextShopID();
+            $contextShopId = (int) $contextShopId > 0 ? $contextShopId : null;
+
+            $shopConstraint = new ShopConstraint(
+                $contextShopId,
+                $contextShopGroup->id
+            );
+        }
+
+        if (isset($configuration['enable_shop'])) {
+            $this->configuration->set('PS_SHOP_ENABLE', $configuration['enable_shop'], $shopConstraint);
+        }
+        if (isset($configuration['maintenance_ip'])) {
+            $this->configuration->set('PS_MAINTENANCE_IP', $configuration['maintenance_ip'], $shopConstraint);
+        }
+        if (isset($configuration['maintenance_text'])) {
+            $this->configuration->set('PS_MAINTENANCE_TEXT', $configuration['maintenance_text'], $shopConstraint, ['html' => true]);
         }
 
         return [];
     }
 
     /**
-     * {@inheritdoc}
+     * @param array $configuration
+     *
+     * @return array
+     */
+    private function removeDisabledFields(array $configuration): array
```

Comment:
I remember that on legacy `AdminController` we could inject our own code `before` and `after` update of the configuration, maybe instead of having this here we could have those `before` and `after` methods and override `removeDisabledFields` only if it's necessary? 

If you think that's not a good idea then I'd see this as a contract with dedicated interface such as `MultiStoreConfiguratorInterface` (as not everyone wants to use those checkboxes or multi-store capabilities in configuration form at all)

Reply:
done !